File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def help_function(self, root:TreeNode, maks: int, good_counter: int):
        if root == None:
            return good_counter

        if maks  <= root.val:
            print("we are here: ", root.val)
            good_counter+=1
            maks = root.val
        
        good_counter = self.help_function(root.left, maks, good_counter)
        good_counter = self.help_function(root.right, maks, good_counter)
        return good_counter

    def goodNodes(self, root: TreeNode) -> int:
        return self.help_function(root, -101, 0)

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int help_function(TreeNode* root, int& k, int& counter)
    {
        if(root == nullptr)
        {
            cout << "we are inside nullptr "<<'\n';;
            if(k == counter)
            {
                k--;
            }
            return -1;
        }
        cout << "hi we are here: " << root->val<<'\n';;
        int num = help_function(root->left, k, counter);
        if(num != -1)return num;

        cout << "back" << '\n';
        if(counter==0)return root->val;

        if(counter != k)
        {
            cout << "hi we are moving up on: " << root->val<<'\n';
            k--;
        }

        num = help_function(root->right, k, counter);
        if(num != -1)return num;
        return -1;
    }

    int kthSmallest(TreeNode* root, int k) {
        return help_function(root, k, k);
    }
};
